### PR TITLE
Batch Requests with System.Text.Json

### PR DESCRIPTION
**Pull request by andrueastman**
### Changes proposed in this PR
This pull request introduces the use of System.Text.Json for marshalling and unmarshalling of batch requests and therefore the change involves the replacement of Newtonsoft in the batch request code.

- As System.Text.Json still does not have a writable Json Dom(equivalents for JArray,JObject etc) the code involves the use of the Utf8JsonWriter to write out json in instances where json writing is needed.

- The previous tests that existed are unbroken by the change. This therefore validates that the functionality remains the same with the added benefits of System.Text.Json. 😄 

### Usage (Unchanged)
Making of batch requests remains the same and can be done as before as follows: - 

```csharp
// Create http GET request.
HttpRequestMessage httpRequestMessage1 = new HttpRequestMessage(HttpMethod.Get, "https://graph.microsoft.com/v1.0/me/");

// Create http POST request.
String jsonContent = "{" +
            "\"displayName\": \"My Notebook\"" +
            "}";
HttpRequestMessage httpRequestMessage2 = new HttpRequestMessage(HttpMethod.Post, "https://graph.microsoft.com/v1.0/me/onenote/notebooks");
httpRequestMessage2.Content = new StringContent(jsonContent, Encoding.UTF8, "application/json");

// Create batch request steps with request ids.
BatchRequestStep requestStep1 = new BatchRequestStep("1", httpRequestMessage1, null);
BatchRequestStep requestStep2 = new BatchRequestStep("2", httpRequestMessage2, new List<string> { "1" });

// Add batch request steps to BatchRequestContent.
BatchRequestContent batchRequestContent = new BatchRequestContent();
batchRequestContent.AddBatchRequestStep(requestStep1);
batchRequestContent.AddBatchRequestStep(requestStep2);

// Send batch request with BatchRequestContent.
HttpResponseMessage response = await httpClient.PostAsync("https://graph.microsoft.com/v1.0/$batch", batchRequestContent);

// Handle http responses using BatchResponseContent.
BatchResponseContent batchResponseContent = new BatchResponseContent(response);
Dictionary<string, HttpResponseMessage> responses = await batchResponseContent.GetResponsesAsync();
HttpResponseMessage httpResponse = await batchResponseContent.GetResponseByIdAsync("1");
string nextLink = await batchResponseContent.GetNextLinkAsync();
```

### Other links
- [Batch Request Content SDK Design](https://github.com/microsoftgraph/msgraph-sdk-design/blob/master/content/BatchRequestContent.md)
- [Batch Response Content SDK Design](https://github.com/microsoftgraph/msgraph-sdk-design/blob/master/content/BatchResponseContent.md)

### CODE REVIEW

File: src/Microsoft.Graph.Core/Requests/BatchRequest.cs [hunk 1]
```diff
@@ -49,9 +49,11 @@ public async Task<BatchResponseContent> PostAsync(BatchRequestContent batchReque
         {
             this.ContentType = "application/json";
             this.Method = "POST";
-            JObject serializableContent = await batchRequestContent.GetBatchRequestContentAsync().ConfigureAwait(false);
-            HttpResponseMessage httpResponseMessage = await this.SendRequestAsync(serializableContent, cancellationToken).ConfigureAwait(false);
-            return new BatchResponseContent(httpResponseMessage);
+            using (Stream serializableContent = await batchRequestContent.GetBatchRequestContentAsync().ConfigureAwait(false))
```
Comment (ddyett) [inline src/Microsoft.Graph.Core/Requests/BatchRequest.cs]: >Stream [](start = 19, length = 6)

STYLE: there's some inconsistency in the usage of var or not.

File: src/Microsoft.Graph.Core/Requests/Content/BatchRequestContent.cs [hunk 2]
```diff
@@ -176,55 +175,93 @@ public bool RemoveBatchRequestStepWithId(string requestId)
...
+            }
 
-            if(batchRequestStep.Request != null && batchRequestStep.Request.Content != null)
+            // write any headers the step contains
+            if (batchRequestStep.Request.Content?.Headers != null && batchRequestStep.Request.Content.Headers.Any())
             {
-                jRequestContent.Add(CoreConstants.BatchRequest.Body, await GetRequestContentAsync(batchRequestStep.Request));
+                writer.WritePropertyName(CoreConstants.BatchRequest.Headers);
+                writer.WriteStartObject();
+                foreach (KeyValuePair<string, IEnumerable<string>> header in batchRequestStep.Request.Content.Headers)
```
Comment (ddyett) [inline src/Microsoft.Graph.Core/Requests/Content/BatchRequestContent.cs]: >KeyValuePair<string, IEnumerable<string>> [](start = 25, length = 41)

var

File: src/Microsoft.Graph.Core/Microsoft.Graph.Core.csproj [hunk 3]
```diff
@@ -86,5 +86,6 @@
     <PackageReference Include="Newtonsoft.Json" Version="9.0.1" />
```
Comment (ddyett) [inline src/Microsoft.Graph.Core/Microsoft.Graph.Core.csproj]: ><PackageReference Include="Newtonsoft.Json" Version="9.0.1" /> [](start = 4, length = 62)

is this still needed?

File: tests/Microsoft.Graph.DotnetCore.Core.Test/Requests/Content/BatchRequestContentTests.cs [hunk 4]
```diff
@@ -189,8 +194,10 @@ public async System.Threading.Tasks.Task BatchRequestContent_GetBatchRequestCont
         [Fact]
         public async System.Threading.Tasks.Task BatchRequestContent_GetBatchRequestContentFromStepAsyncDoesNotModifyDateTimes()
         {
+            // System.Text.Json is strict on json content by default. So make sure that there are no 
+            // trailing comma's and special characters
             var payloadString = "{\r\n" +
-                                "  \"subject\": \"Let's go for lunch\",\r\n" +
+                                "  \"subject\": \"Lets go for lunch\",\r\n" +
```
Comment (ddyett) [inline tests/Microsoft.Graph.DotnetCore.Core.Test/Requests/Content/BatchRequestContentTests.cs]: What if someone was passing it in as "Let's".  Is there code now broken?

File: tests/Microsoft.Graph.DotnetCore.Core.Test/Requests/Content/BatchResponseContentTests.cs [hunk 5]
```diff
@@ -30,7 +30,7 @@ public async Task BatchResponseContent_InitializeWithNoContentAsync()
         [Fact]
         public async Task BatchResponseContent_InitializeWithEmptyResponseContentAsync()
         {
-            string jsonResponse = "{ responses: [] }";
+            string jsonResponse = "{ \"responses\": [] }";
```
Comment (ddyett) [inline tests/Microsoft.Graph.DotnetCore.Core.Test/Requests/Content/BatchResponseContentTests.cs]: >\"responses\" [](start = 37, length = 13)

was this a prior issue with the test or something you had to do with the change.  What I'm getting it as that I would have expected test content not to need a change if all things were equal.

File: src/Microsoft.Graph.Core/Microsoft.Graph.Core.csproj (hunk 3 above)
Comment (andrueastman) [reply src/Microsoft.Graph.Core/Microsoft.Graph.Core.csproj]: Yes. There is still code using Newtonsoft. Which will be changed in upcoming PRs.

File: tests/Microsoft.Graph.DotnetCore.Core.Test/Requests/Content/BatchRequestContentTests.cs (hunk 4 above)
Comment (andrueastman) [reply tests/Microsoft.Graph.DotnetCore.Core.Test/Requests/Content/BatchRequestContentTests.cs]: No.The serializet will however encode the the `'` character. I only removed it to simplify the test.

File: tests/Microsoft.Graph.DotnetCore.Core.Test/Requests/Content/BatchResponseContentTests.cs (hunk 5 above)
Comment (andrueastman) [reply tests/Microsoft.Graph.DotnetCore.Core.Test/Requests/Content/BatchResponseContentTests.cs]: System.text.json is more strict with the json standard. Therefore, properties without the double quotes are not valid json.

File: src/Microsoft.Graph.Core/Microsoft.Graph.Core.csproj [hunk 6]
```diff
@@ -86,5 +86,6 @@
...
     <PackageReference Include="Microsoft.Identity.Client" Version="4.7.1" />
```
[the remark this answers is not in the document]
Comment (andrueastman) [reply src/Microsoft.Graph.Core/Microsoft.Graph.Core.csproj]: Thanks for pointing this out. This should now be resolved.

File: tests/Microsoft.Graph.DotnetCore.Core.Test/Requests/Content/BatchRequestContentTests.cs [hunk 7]
```diff
@@ -261,7 +273,7 @@ public async System.Threading.Tasks.Task BatchRequestContent_GetBatchRequestCont
                                   "          \"timeZone\": \"Pacific Standard Time\"\r\n" +
                                   "        },\r\n" +
                                   "        \"location\": {\r\n" +
-                                  "          \"displayName\": \"Harry's Bar\"\r\n" +
+                                  "          \"displayName\": \"Harrys Bar\"\r\n" +
```
Comment (darrelmiller) [inline tests/Microsoft.Graph.DotnetCore.Core.Test/Requests/Content/BatchRequestContentTests.cs]: Is there no escape character for single ticks?

Comment (andrueastman) [reply tests/Microsoft.Graph.DotnetCore.Core.Test/Requests/Content/BatchRequestContentTests.cs]: More info is here https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-migrate-from-newtonsoft-how-to#minimal-character-escaping.

The single tick character will be escaped to emit `\u0027`

**Pull request merged**